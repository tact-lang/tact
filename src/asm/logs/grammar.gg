vmLine =
      VmLoc
    / VmStack
    / VmExecute
    / VmLimitChanged
    / VmGasRemaining
    / VmException
    / VmExceptionHandler
    / VmFinalC5
    / VmUnknown;

VmStack            = "stack: " stack:$[^\n]+;
VmLoc              = "code cell hash:" hash:hex "offset:" offset:number;
VmExecute          = "execute " instr:$[^\n]+;
VmLimitChanged     = "changing gas limit to " limit:number;
VmGasRemaining     = "gas remaining: " gas:number;
VmException        = "handling exception code " errno:number ": " message:$[^\n]*;
VmExceptionHandler = "default exception handler, terminating vm with exit code" errno:number;
VmFinalC5          = "final c5:" value:Cell;
VmUnknown          = !"stack" text:$[^\n]+ "\n";

VmParsedStack = "[" values:VmStackValue* "]";

VmStackValue = value:(
      Null
    / NaN
    / Integer
    / Tuple
    / TupleParen
    / Cell
    / Continuation
    / Builder
    / CellSlice
    / Unknown
);

Null         = "()" / "(null)";
NaN          = "NaN";
Integer      = value:number;
Tuple        = "[" elements:VmStackValue* "]";
TupleParen   = "(" elements:VmStackValue* ")";
Cell         = "C{" value:hex "}";
Continuation = "Cont{" value:$#([A-Za-z_0-9]*) "}";
Builder      = "BC{" value:hex "}";
Unknown      = "???";

CellSlice     = "CS{" body:(CellSliceBody / CellSliceShortBody) "}";
CellSliceBody = "Cell{" value:hex "}" bits:CellSliceBits ";" refs:CellSliceRefs;
CellSliceBits = "bits:" start:number ".." end:number;
CellSliceRefs = "refs:" start:number ".." end:number;

CellSliceShortBody = value:hex;

number   = op:"-"? value:#$(digit+);
digit    = [0-9];

hexDigit "hexadecimal digit" = [0-9a-fA-F];
hex      = $#(value:hexDigit*);

space "space" = [ \t\r\n];
