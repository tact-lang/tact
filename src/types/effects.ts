import type { AstExpression, AstId, AstStatement } from "../ast/ast";
import { idText, isSelfId, tryExtractPath } from "../ast/ast-helpers";
import type { CompilerContext } from "../context/context";
import { getAllTypes, getType } from "./resolveDescriptors";
import { idTextErr, throwInternalCompilerError } from "../error/errors";
import { getExpType } from "./resolveExpression";

export type Effect = "contractStorageRead" | "contractStorageWrite";

export function computeReceiversEffects(ctx: CompilerContext) {
    for (const type of getAllTypes(ctx)) {
        if (type.kind === "contract") {
            for (const receiver of type.receivers) {
                receiver.effects = statementListEffects(
                    receiver.ast.statements,
                    new Set<AstId>(),
                    ctx,
                );
            }
        }
    }
}

function statementListEffects(
    statements: readonly AstStatement[],
    processedContractMethods: ReadonlySet<AstId>,
    ctx: CompilerContext,
): ReadonlySet<Effect> {
    return mapUnionAll(statements, (stmt) =>
        statementEffects(stmt, processedContractMethods, ctx),
    );
}

function statementEffects(
    stmt: AstStatement,
    processedContractMethods: ReadonlySet<AstId>,
    ctx: CompilerContext,
): ReadonlySet<Effect> {
    switch (stmt.kind) {
        case "statement_let":
        case "statement_expression":
        case "statement_destruct": {
            return expressionEffects(
                stmt.expression,
                processedContractMethods,
                ctx,
            );
        }
        case "statement_return": {
            return stmt.expression
                ? expressionEffects(
                      stmt.expression,
                      processedContractMethods,
                      ctx,
                  )
                : new Set<Effect>();
        }
        case "statement_assign":
        case "statement_augmentedassign": {
            // since we don't analyze method bodies, we know `self` refers to the contract itself
            const [head, _] = tryExtractPath(stmt.path) ?? [];
            if (typeof head === "undefined") {
                throwInternalCompilerError(
                    "Not an l-value and typechecker should have caught it",
                    stmt.path.loc,
                );
            }

            const rhsEffects = expressionEffects(
                stmt.expression,
                processedContractMethods,
                ctx,
            );
            if (isSelfId(head)) {
                const lvalueEffects =
                    stmt.kind === "statement_augmentedassign"
                        ? new Set<Effect>([
                              "contractStorageWrite",
                              "contractStorageRead",
                          ])
                        : new Set<Effect>(["contractStorageWrite"]);
                return union(lvalueEffects, rhsEffects);
            } else {
                return rhsEffects;
            }
        }
        case "statement_condition": {
            const conditionEffects = expressionEffects(
                stmt.condition,
                processedContractMethods,
                ctx,
            );
            const thenEffects = statementListEffects(
                stmt.trueStatements,
                processedContractMethods,
                ctx,
            );
            const elseEffects = statementListEffects(
                stmt.falseStatements ?? [],
                processedContractMethods,
                ctx,
            );
            return unionAll([conditionEffects, thenEffects, elseEffects]);
        }
        case "statement_while":
        case "statement_until": {
            const conditionEffects = expressionEffects(
                stmt.condition,
                processedContractMethods,
                ctx,
            );
            const bodyEffects = statementListEffects(
                stmt.statements,
                processedContractMethods,
                ctx,
            );
            return union(conditionEffects, bodyEffects);
        }
        case "statement_repeat": {
            const iterationsEffects = expressionEffects(
                stmt.iterations,
                processedContractMethods,
                ctx,
            );
            const bodyEffects = statementListEffects(
                stmt.statements,
                processedContractMethods,
                ctx,
            );
            return union(iterationsEffects, bodyEffects);
        }
        case "statement_try": {
            const tryEffects = statementListEffects(
                stmt.statements,
                processedContractMethods,
                ctx,
            );
            const catchEffects = statementListEffects(
                stmt.catchBlock?.catchStatements ?? [],
                processedContractMethods,
                ctx,
            );
            return union(tryEffects, catchEffects);
        }
        case "statement_foreach": {
            const mapExpressionEffects = expressionEffects(
                stmt.map,
                processedContractMethods,
                ctx,
            );
            const bodyEffects = statementListEffects(
                stmt.statements,
                processedContractMethods,
                ctx,
            );
            return union(mapExpressionEffects, bodyEffects);
        }
        case "statement_block": {
            return statementListEffects(
                stmt.statements,
                processedContractMethods,
                ctx,
            );
        }
    }
}

function expressionEffects(
    expr: AstExpression,
    processedContractMethods: ReadonlySet<AstId>,
    ctx: CompilerContext,
): ReadonlySet<Effect> {
    switch (expr.kind) {
        case "id": {
            return new Set<Effect>();
        }
        case "field_access": {
            // we only analyze receiver bodies and contract methods calls
            // so we now `self` refers to the contract itself
            return expr.aggregate.kind === "id" && isSelfId(expr.aggregate)
                ? new Set<Effect>(["contractStorageRead"])
                : expressionEffects(
                      expr.aggregate,
                      processedContractMethods,
                      ctx,
                  );
        }
        case "method_call": {
            const argsEffects = mapUnionAll(expr.args, (arg) =>
                expressionEffects(arg, processedContractMethods, ctx),
            );
            const selfEffects = expressionEffects(
                expr.self,
                processedContractMethods,
                ctx,
            );
            const methodCallEffects = methodEffects(
                expr.self,
                expr.method,
                processedContractMethods,
                ctx,
            );

            return unionAll([argsEffects, selfEffects, methodCallEffects]);
        }
        case "string":
        case "number":
        case "boolean":
        case "slice":
        case "null":
        case "simplified_string":
        case "address":
        case "cell":
        case "struct_value":
        case "code_of": {
            return new Set<Effect>();
        }
        case "op_binary": {
            const leftEffects = expressionEffects(
                expr.left,
                processedContractMethods,
                ctx,
            );
            const rightEffects = expressionEffects(
                expr.right,
                processedContractMethods,
                ctx,
            );
            return union(leftEffects, rightEffects);
        }
        case "op_unary": {
            return expressionEffects(
                expr.operand,
                processedContractMethods,
                ctx,
            );
        }
        case "conditional": {
            const conditionEffects = expressionEffects(
                expr.condition,
                processedContractMethods,
                ctx,
            );
            const thenEffects = expressionEffects(
                expr.thenBranch,
                processedContractMethods,
                ctx,
            );
            const elseEffects = expressionEffects(
                expr.elseBranch,
                processedContractMethods,
                ctx,
            );
            return unionAll([conditionEffects, thenEffects, elseEffects]);
        }
        case "init_of":
        case "static_call": {
            // global (static) functions cannot change contract storage because of the call-by-value semantics, so we don't analyze their bodies
            return mapUnionAll(expr.args, (arg) =>
                expressionEffects(arg, processedContractMethods, ctx),
            );
        }
        case "struct_instance": {
            return mapUnionAll(expr.args, (field) =>
                expressionEffects(
                    field.initializer,
                    processedContractMethods,
                    ctx,
                ),
            );
        }
    }
}

function methodEffects(
    self: AstExpression,
    method: AstId,
    processedContractMethods: ReadonlySet<AstId>,
    ctx: CompilerContext,
): ReadonlySet<Effect> {
    const selfTypeRef = getExpType(ctx, self);

    // contract method call: self.foo(), since variable shadowing is not allowed
    if (selfTypeRef.kind === "ref") {
        const selfType = getType(ctx, selfTypeRef.name);

        if (selfType.kind === "contract") {
            if (processedContractMethods.has(method)) {
                return new Set();
            }

            const methodDescr = selfType.functions.get(idText(method));
            if (typeof methodDescr === "undefined") {
                throwInternalCompilerError(
                    `Method ${idTextErr(method)} not found in contract ${selfTypeRef.name}`,
                    method.loc,
                );
            }
            switch (methodDescr.ast.kind) {
                case "function_decl":
                    {
                        throwInternalCompilerError(
                            `Cannot call a function declaration ${idTextErr(method)} on contract ${selfTypeRef.name}`,
                            method.loc,
                        );
                    }
                    break;
                case "asm_function_def":
                case "native_function_decl": {
                    // Cannot analyze the effects of native and asm functions
                    // so we make the most conservative approximation
                    return methodDescr.isMutating
                        ? new Set<Effect>([
                              "contractStorageRead",
                              "contractStorageWrite",
                          ])
                        : new Set<Effect>(["contractStorageRead"]);
                }
                case "function_def": {
                    // essentially we inline contract method calls (modulo recursion)
                    return statementListEffects(
                        methodDescr.ast.statements,
                        addToSet(processedContractMethods, method),
                        ctx,
                    );
                }
            }
        }
    }

    const [head, ...rest] = tryExtractPath(self) ?? [];

    // method call on a contract storage variable: e.g. self.x.inc()
    if (typeof head !== "undefined" && isSelfId(head) && rest.length > 0) {
        switch (selfTypeRef.kind) {
            case "map":
                {
                    switch (idText(method)) {
                        case "set":
                        case "replace":
                        case "replaceGet":
                        case "del": {
                            return new Set<Effect>([
                                "contractStorageRead",
                                "contractStorageWrite",
                            ]);
                        }
                        case "get":
                        case "asCell":
                        case "isEmpty":
                        case "exists":
                        case "deepEquals": {
                            return new Set<Effect>(["contractStorageRead"]);
                        }
                        default:
                            throwInternalCompilerError(
                                `Invalid method call on map: ${idTextErr(method)}`,
                                method.loc,
                            );
                    }
                }
                break;
            case "ref": {
                const selfType = getType(ctx, selfTypeRef.name);
                const methodDescr = selfType.functions.get(idText(method));
                if (typeof methodDescr === "undefined") {
                    throwInternalCompilerError(
                        `Method ${idTextErr(method)} not found in type ${selfTypeRef.name}`,
                        method.loc,
                    );
                }
                return methodDescr.isMutating
                    ? new Set<Effect>([
                          "contractStorageRead",
                          "contractStorageWrite",
                      ])
                    : new Set<Effect>(["contractStorageRead"]);
            }
            case "ref_bounced": // Extend functions must have a reference type as the first parameter
            case "void":
            case "null": {
                throwInternalCompilerError(
                    "Invalid type for method call",
                    self.loc,
                );
            }
        }
    }
    return new Set<Effect>();
}

function addToSet<T>(set: ReadonlySet<T>, element: T): ReadonlySet<T> {
    return new Set([...set, element]);
}

function union<T>(left: ReadonlySet<T>, right: ReadonlySet<T>): ReadonlySet<T> {
    return new Set([...left, ...right]);
}

function unionAll<T>(sets: readonly ReadonlySet<T>[]): ReadonlySet<T> {
    return new Set(sets.flatMap((set) => [...set]));
}

function mapUnionAll<S, T>(
    xs: readonly S[],
    f: (x: S) => ReadonlySet<T>,
): ReadonlySet<T> {
    return xs.reduce<ReadonlySet<T>>(
        (acc, x) => union(acc, f(x)),
        new Set<T>(),
    );
}
