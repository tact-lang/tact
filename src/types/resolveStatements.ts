import { CompilerContext } from "../context";
import { ASTCondition, ASTRef, ASTStatement, throwError } from "../grammar/ast";
import { isAssignable } from "./isAssignable";
import { getAllStaticFunctions, getAllTypes, resolveTypeRef } from "./resolveDescriptors";
import { getExpType, resolveExpression, resolveLValueRef } from "./resolveExpression";
import { printTypeRef, TypeRef } from "./types";

export type StatementContext = {
    root: ASTRef,
    returns: TypeRef,
    vars: { [name: string]: TypeRef };
    requiredFields: string[];
};

function emptyContext(root: ASTRef, returns: TypeRef): StatementContext {
    return {
        root,
        returns,
        vars: {},
        requiredFields: []
    };
}

function addRequiredVariables(name: string, src: StatementContext): StatementContext {
    if (src.requiredFields.find((v) => v === name)) {
        throw Error('Variable already exists: ' + name); // Should happen earlier
    }
    return {
        ...src,
        requiredFields: [
            ...src.requiredFields,
            name
        ]
    };
}

function removeRequiredVariable(name: string, src: StatementContext): StatementContext {
    if (!src.requiredFields.find((v) => v === name)) {
        throw Error('Variable is not required: ' + name); // Should happen earlier
    }
    const filtered = src.requiredFields.filter((v) => v !== name);
    return {
        ...src,
        requiredFields: filtered
    };
}

function addVariable(name: string, ref: TypeRef, src: StatementContext): StatementContext {
    if (src.vars[name]) {
        throw Error('Variable already exists: ' + name); // Should happen earlier
    }
    return {
        ...src,
        vars: {
            ...src.vars,
            [name]: ref
        }
    };
}

function processCondition(condition: ASTCondition, sctx: StatementContext, ctx: CompilerContext): { ctx: CompilerContext, sctx: StatementContext } {

    // Process expression
    ctx = resolveExpression(condition.expression, sctx, ctx);
    let initialCtx = sctx;

    // Simple if
    if (condition.falseStatements === null && condition.elseif === null) {
        const r = processStatements(condition.trueStatements, initialCtx, ctx);
        ctx = r.ctx;
        return { ctx, sctx: initialCtx };
    }

    // Simple if-else
    const processedCtx: StatementContext[] = [];

    // Process true branch
    const r = processStatements(condition.trueStatements, initialCtx, ctx);
    ctx = r.ctx;
    processedCtx.push(r.sctx);

    // Process else/elseif branch
    if (condition.falseStatements !== null && condition.elseif === null) {
        // if-else
        const r = processStatements(condition.falseStatements, initialCtx, ctx);
        ctx = r.ctx;
        processedCtx.push(r.sctx);
    } else if (condition.falseStatements === null && condition.elseif !== null) {
        // if-else if
        const r = processCondition(condition.elseif, initialCtx, ctx);
        ctx = r.ctx;
        processedCtx.push(r.sctx);
    } else {
        throw Error('Impossible');
    }

    // Merge statement contexts
    const removed: string[] = [];
    for (const f of initialCtx.requiredFields) {
        let found = false;
        for (const c of processedCtx) {
            if (c.requiredFields.find((v) => v === f)) {
                found = true;
                break;
            }
        }
        if (!found) {
            removed.push(f);
        }
    }
    for (const r of removed) {
        initialCtx = removeRequiredVariable(r, initialCtx);
    }

    return { ctx, sctx: initialCtx };
}

function processStatements(statements: ASTStatement[], sctx: StatementContext, ctx: CompilerContext): { ctx: CompilerContext, sctx: StatementContext } {

    // Process statements

    let exited = false;
    for (const s of statements) {

        // Check for unreachable
        if (exited) {
            throwError('Unreachable statement', s.ref);
        }

        // Process statement
        if (s.kind === 'statement_let') {

            // Process expression
            ctx = resolveExpression(s.expression, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.expression);
            const variableType = resolveTypeRef(ctx, s.type);
            if (!isAssignable(expressionType, variableType)) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to ${printTypeRef(variableType)}`, s.ref);
            }

            // Add variable to statement context
            if (sctx.vars[s.name]) {
                throwError(`Variable already exists: ${s.name}`, s.ref);
            }
            sctx = addVariable(s.name, variableType, sctx);

        } else if (s.kind === 'statement_assign') {

            // Process lvalue
            ctx = resolveLValueRef(s.path, sctx, ctx);

            // Process expression
            ctx = resolveExpression(s.expression, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.expression);
            const tailType = getExpType(ctx, s.path[s.path.length - 1]);
            if (!isAssignable(expressionType, tailType)) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to ${printTypeRef(tailType)}`, s.ref);
            }

            // Mark as assigned
            if (s.path.length === 2 && s.path[0].name === 'self') {
                const field = s.path[1].name;
                if (sctx.requiredFields.findIndex((v) => v === field) >= 0) {
                    sctx = removeRequiredVariable(field, sctx);
                }
            }

        } else if (s.kind == 'statement_augmentedassign') {
                
            // Process lvalue
            ctx = resolveLValueRef(s.path, sctx, ctx);

            // Process expression
            ctx = resolveExpression(s.expression, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.expression);
            const tailType = getExpType(ctx, s.path[s.path.length - 1]);
            if (!isAssignable(expressionType, tailType)) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to ${printTypeRef(tailType)}`, s.ref);
            }

            // Mark as assigned
            if (s.path.length === 2 && s.path[0].name === 'self') {
                const field = s.path[1].name;
                if (sctx.requiredFields.findIndex((v) => v === field) >= 0) {
                    sctx = removeRequiredVariable(field, sctx);
                }
            }

        } else if (s.kind === 'statement_expression') {

            // Process expression
            ctx = resolveExpression(s.expression, sctx, ctx);

        } else if (s.kind === 'statement_condition') {

            // Process condition (expression resolved inside)
            const r = processCondition(s, sctx, ctx);
            ctx = r.ctx;
            sctx = r.sctx;

            // Check type
            const expressionType = getExpType(ctx, s.expression);
            if (expressionType.kind !== 'ref' || expressionType.name !== 'Bool' || expressionType.optional) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to Bool`, s.ref);
            }

        } else if (s.kind === 'statement_return') {

            if (s.expression) {

                // Process expression
                ctx = resolveExpression(s.expression, sctx, ctx);

                // Check type
                const expressionType = getExpType(ctx, s.expression);
                if (!isAssignable(expressionType, sctx.returns)) {
                    throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to ${printTypeRef(sctx.returns)}`, s.ref);
                }
            } else {
                if (sctx.returns.kind !== 'void') {
                    throwError(`Type mismatch: void is not assignable to ${printTypeRef(sctx.returns)}`, s.ref);
                }
            }

            // Check if all required variables are assigned
            if (sctx.requiredFields.length > 0) {
                if (sctx.requiredFields.length === 1) {
                    throwError(`Field ${sctx.requiredFields[0]} is not set`, sctx.root);
                } else {
                    throwError(`Fields ${sctx.requiredFields.join(', ')} are not set`, sctx.root);
                }
            }

            // Mark as ended
            exited = true;

        } else if (s.kind === 'statement_repeat') {

            // Process expression
            ctx = resolveExpression(s.condition, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.condition);
            if (expressionType.kind !== 'ref' || expressionType.name !== 'Int' || expressionType.optional) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to Int`, s.ref);
            }

            // Process inner statements
            const r = processStatements(s.statements, sctx, ctx);
            ctx = r.ctx;
            sctx = r.sctx;

        } else if (s.kind === 'statement_until') {

            // Process expression
            ctx = resolveExpression(s.condition, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.condition);
            if (expressionType.kind !== 'ref' || expressionType.name !== 'Bool' || expressionType.optional) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to bool`, s.ref);
            }

            // Process inner statements
            const r = processStatements(s.statements, sctx, ctx);
            ctx = r.ctx;
            sctx = r.sctx;

        } else if (s.kind === 'statement_while') {

            // Process expression
            ctx = resolveExpression(s.condition, sctx, ctx);

            // Check type
            const expressionType = getExpType(ctx, s.condition);
            if (expressionType.kind !== 'ref' || expressionType.name !== 'Bool' || expressionType.optional) {
                throwError(`Type mismatch: ${printTypeRef(expressionType)} is not assignable to bool`, s.ref);
            }

            // Process inner statements
            const r = processStatements(s.statements, sctx, ctx);
            ctx = r.ctx;
            sctx = r.sctx;

        } else {
            throw Error('Unknown statement');
        }
    }

    return { ctx, sctx };
}

function processFunctionBody(statements: ASTStatement[], sctx: StatementContext, ctx: CompilerContext) {
    const res = processStatements(statements, sctx, ctx);

    // Check if all required variables are assigned
    if (res.sctx.requiredFields.length > 0) {
        if (res.sctx.requiredFields.length === 1) {
            throwError(`Field ${res.sctx.requiredFields[0]} is not set`, res.sctx.root);
        } else {
            throwError(`Fields ${res.sctx.requiredFields.join(', ')} are not set`, res.sctx.root);
        }
    }

    return res.ctx;
}

export function resolveStatements(ctx: CompilerContext) {

    // Process all static functions
    for (const f of Object.values(getAllStaticFunctions(ctx))) {
        if (f.ast.kind === 'def_function') {

            // Build statement context
            let sctx = emptyContext(f.ast.ref, f.returns);
            for (const p of f.args) {
                sctx = addVariable(p.name, p.type, sctx);
            }

            // Process
            if (f.ast.statements) {
                ctx = processFunctionBody(f.ast.statements, sctx, ctx);
            }
        }
    }

    // Process all types
    for (const t of Object.values(getAllTypes(ctx))) {

        // Process init
        if (t.init) {

            // Build statement context
            let sctx = emptyContext(t.init.ast.ref, { kind: 'void' });

            // Self
            sctx = addVariable('self', { kind: 'ref', name: t.name, optional: false }, sctx);

            // Required variables
            for (const f of t.fields) {
                if (f.default !== undefined) { // NOTE: undefined is important here
                    continue;
                }
                if (isAssignable({ kind: 'null' }, f.type)) {
                    continue;
                }
                sctx = addRequiredVariables(f.name, sctx);
            }

            // Args
            for (const p of t.init.args) {
                sctx = addVariable(p.name, p.type, sctx);
            }

            // Process
            ctx = processFunctionBody(t.init.ast.statements, sctx, ctx);
        }

        // Process receivers
        for (const f of Object.values(t.receivers)) {

            // Build statement context
            let sctx = emptyContext(f.ast.ref, { kind: 'void' });
            sctx = addVariable('self', { kind: 'ref', name: t.name, optional: false }, sctx);
            if (f.selector.kind === 'internal-binary' || f.selector.kind === 'external-binary') {
                sctx = addVariable(f.selector.name, { kind: 'ref', name: f.selector.type, optional: false }, sctx);
            } else if (f.selector.kind === 'internal-empty' || f.selector.kind === 'external-empty' || f.selector.kind === 'external-comment' || f.selector.kind === 'internal-comment') {
                // Nothing to add to context
            } else if (f.selector.kind === 'internal-comment-fallback' || f.selector.kind === 'external-comment-fallback') {
                sctx = addVariable(f.selector.name, { kind: 'ref', name: 'String', optional: false }, sctx);
            } else if (f.selector.kind === 'internal-fallback' || f.selector.kind === 'external-fallback') {
                sctx = addVariable(f.selector.name, { kind: 'ref', name: 'Slice', optional: false }, sctx);
            } else if (f.selector.kind === 'bounce-fallback') {
                sctx = addVariable(f.selector.name, { kind: 'ref', name: 'Slice', optional: false }, sctx);
            } else if (f.selector.kind === 'bounce-binary') {
                sctx = addVariable(f.selector.name, f.selector.bounced ? { kind: 'ref_bounced', name: f.selector.type } : { kind: 'ref', name: f.selector.type, optional: false }, sctx);
            } else {
                throw Error('Unknown selector');
            }

            // Process
            ctx = processFunctionBody(f.ast.statements, sctx, ctx);
        }

        // Process functions
        for (const f of t.functions.values()) {
            if (f.ast.kind !== 'def_native_function') {

                // Build statement context
                let sctx = emptyContext(f.ast.ref, f.returns);
                sctx = addVariable('self', { kind: 'ref', name: t.name, optional: false }, sctx);
                for (const a of f.args) {
                    sctx = addVariable(a.name, a.type, sctx);
                }

                // Process
                if (f.ast.statements) {
                    ctx = processFunctionBody(f.ast.statements, sctx, ctx);
                }
            }
        }
    }

    return ctx;
}