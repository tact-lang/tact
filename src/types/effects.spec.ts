import { getAllTypes, resolveDescriptors } from "./resolveDescriptors";
import { loadCases } from "../utils/loadCases";
import { openContext } from "../context/store";
import { resolveStatements } from "./resolveStatements";
import { CompilerContext } from "../context/context";
import { featureEnable } from "../config/features";
import { getParser } from "../grammar";
import { getAstFactory } from "../ast/ast-helpers";
import { defaultParser } from "../grammar/grammar";
import { computeReceiversEffects } from "./effects";

describe("effects", () => {
    for (const testContract of loadCases(__dirname + "/effects/")) {
        it(`should correctly compute effects: ${testContract.name}`, () => {
            const Ast = getAstFactory();
            let ctx = openContext(
                new CompilerContext(),
                [
                    {
                        code: testContract.code,
                        path: "<unknown>",
                        origin: "user",
                    },
                ],
                [],
                getParser(Ast, defaultParser),
            );
            ctx = featureEnable(ctx, "external");
            ctx = resolveDescriptors(ctx, Ast);
            ctx = resolveStatements(ctx, Ast);
            computeReceiversEffects(ctx);
            const receiverEffects = getAllTypes(ctx)
                .filter((type) => type.kind === "contract")
                .map((contract) => {
                    return {
                        contract: contract.name,
                        receivers: contract.receivers.map((receiver) => {
                            return {
                                ...receiver.selector,
                                effects: receiver.effects,
                            };
                        }),
                    };
                });
            expect(receiverEffects).toMatchSnapshot();
        });
    }
});
