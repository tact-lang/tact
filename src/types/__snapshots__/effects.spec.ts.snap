// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effects should correctly compute effects: contractStorageAnalysis 1`] = `
[
  {
    "contract": "StorageAnalysis",
    "receivers": [
      {
        "effects": Set {},
        "kind": "internal-empty",
      },
      {
        "comment": "direct regular assignment into storage var",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "direct regular assignment into storage struct field",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "direct regular assignment: no storage write",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "direct augmented assignment into storage",
        "effects": Set {
          "contractStorageWrite",
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "direct augmented assignment into storage struct field",
        "effects": Set {
          "contractStorageWrite",
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "direct augmented assignment: no storage",
        "effects": Set {},
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a contract variable: mutates",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a contract struct field: mutates",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a contract struct: mutates",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a struct: no mutates - 1",
        "effects": Set {},
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a struct: no mutates - 2",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "calling a mutating method on a contract variable and struct field: no mutation",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "rewrite in a contract method",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "transitively modifying contract method",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "read and modify in a contract method",
        "effects": Set {
          "contractStorageWrite",
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying contract method",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "asm function modifies contract storage via assigning to \`self\`: swapStorage",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "asm function modifies contract storage via assigning to \`self\`: setStorage",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "modifying map method on a contract map",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying map method on a contract map",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "modifying external contract method",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying external contract method",
        "effects": Set {
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying recursive contract method",
        "effects": Set {},
        "kind": "internal-comment",
      },
      {
        "comment": "modifying recursive contract method",
        "effects": Set {
          "contractStorageWrite",
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "modifying recursive external contract method",
        "effects": Set {
          "contractStorageWrite",
          "contractStorageRead",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying mutually recursive contract methods",
        "effects": Set {},
        "kind": "internal-comment",
      },
      {
        "comment": "modifying mutually recursive contract methods",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "modifying getter as a contract method",
        "effects": Set {
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "non-modifying getter as a contract method",
        "effects": Set {},
        "kind": "internal-comment",
      },
      {
        "comment": "complex control flow with deeply nested contract modification",
        "effects": Set {
          "contractStorageRead",
          "contractStorageWrite",
        },
        "kind": "internal-comment",
      },
      {
        "comment": "method call on null",
        "effects": Set {},
        "kind": "internal-comment",
      },
    ],
  },
]
`;
